File: trunk/EngineUAB/Run/Videogame/Data/Scripts/Game/Player/player_target_run_state.lua
class 'CPlayerTargetRunState' (CState)
	function CPlayerTargetRunState:__init() 
		CState.__init(self)
		self.action_2_input = core:get_action_to_input()
	end

	function CPlayerTargetRunState:OnEnter(_CCharacter)
		core:get_debug_gui_manager().debug_render:set_state_name("Target run")
	
		self.enemy_detected = get_game_process():get_character_manager().preview_target_enemy
		if self.enemy_detected == nil then
			self.enemy_detected = _CCharacter:detect_enemy()
		end
		get_game_process():get_character_manager().target_enemy = self.enemy_detected
	end
	
	function CPlayerTargetRunState:Execute(_CCharacter)
		local l_d = 0.0
		local l_yaw = 0.0
		local l_dir = Vect3f(0.0, 0.0, 0.0)
		local l_move_player = false
		
		if not _CCharacter.locked then
			--Mira si se desbloquea el target
			if self.action_2_input:do_action('PlayerTarget') then
				_CCharacter.is_target_fixed = false
			end
			
			--El jugador ataca
			if self.action_2_input:do_action('AttackPlayer') then
				if _CCharacter.is_target_fixed then
					_CCharacter.logic_fsm:change_state(_CCharacter.target_attack)
					_CCharacter.graphic_fsm:change_state(_CCharacter.animated_attack)
				else
					_CCharacter.logic_fsm:change_state(_CCharacter.attack)
					_CCharacter.graphic_fsm:change_state(_CCharacter.animated_attack)
				end
			end
			
			--Calcula el ángulo entre donde miras y el enemigo
			local l_player_dir = get_game_process().player_camera:get_direction()
			local l_enemy_dir = get_game_process():get_character_manager().target_enemy.position - _CCharacter.position
			l_player_dir.y = 0.0
			l_enemy_dir.y = 0.0
			l_enemy_dir:normalize(1.0)
			l_player_dir:normalize(1.0)
			l_d = l_enemy_dir:dot(l_player_dir) --Calcula el ángulo entre donde mira el personaje y la dirección hacia el enemigo		
			l_d = math.acos(l_d)
			
			if (math.deg(l_d) > 0.1)  then
				_CCharacter.yaw = _CCharacter.yaw - l_d
			end
			l_yaw = _CCharacter.yaw
			--Se mueve el personaje con el target asignado
			if self.action_2_input:do_action('MovePlayerUp') then
				l_dir = l_enemy_dir
				l_move_player = true
				
			elseif self.action_2_input:do_action('MovePlayerDown') then
				l_dir = Vect3f(-l_enemy_dir.x, -l_enemy_dir.y, -l_enemy_dir.z)
				l_move_player = true
			
			elseif self.action_2_input:do_action('MovePlayerLeft') then
				local l_vector_yaw = l_yaw + math.pi/2
				l_dir = Vect3f(math.cos(l_vector_yaw), 0.0, math.sin(l_vector_yaw))
				l_move_player = true
				
			elseif self.action_2_input:do_action('MovePlayerRight') then
				_CCharacter.yaw = l_yaw + 2*l_d
				local l_vector_yaw = _CCharacter.yaw - math.pi/2
				l_dir = Vect3f(math.cos(l_vector_yaw), 0.0, math.sin(l_vector_yaw))
				l_move_player = true
			
			else
				l_dir = Vect3f(0.0, 0.0, 0.0)
			end
		end
		
		_CCharacter.yaw = angle_filter(_CCharacter.yaw)
		--Le aplica la velocidad al movimiento
		l_dir = l_dir * 10.0 * _CCharacter.elapsed_time
		
		--Mueve el controller físico
		_CCharacter.physic_controller:move(l_dir, _CCharacter.elapsed_time)
		
		--Pone la rotación del personaje
		_CCharacter.physic_controller.yaw = l_yaw
		_CCharacter.animated_model.yaw = -(l_yaw * 180.0 / math.pi) + 90.0
		
		--Cambia de estado
		if not l_move_player then
			_CCharacter.logic_fsm:change_state(_CCharacter.idle)
			_CCharacter.graphic_fsm:change_state(_CCharacter.animated_idle)
		else
			--Mira si se hace un salto
			if self.action_2_input:do_action('PlayerJump') then 
				_CCharacter.logic_fsm:change_state(_CCharacter.jump)
				_CCharacter.graphic_fsm:change_state(_CCharacter.animated_jump)
			elseif not _CCharacter.is_target_fixed then
				_CCharacter.logic_fsm:change_state(_CCharacter.run)
				_CCharacter.graphic_fsm:change_state(_CCharacter.animated_run)
				get_game_process():get_character_manager().target_enemy = nil
			end
		end
	end
	
	function CPlayerTargetRunState:OnExit(_CCharacter)
	end
	
	function CPlayerTargetRunState:OnMessage(_CCharacter, _Msg)
		if ( _Msg.msg == msg_attack ) then
			_CCharacter.logic_fsm:change_state(_CCharacter.hit)
			_CCharacter.graphic_fsm:change_state(_CCharacter.animated_hit)
			return true
		end
		return false
	end
	
	function CPlayerTargetRunState:__Finalize()
	end

File: EngineUAB/Code/Videogame/Characters/Player/LogicStates/PlayerTargetRunState.cpp
#include "PlayerTargetRunState.h"
#include "StatesMachine\Telegram.h"
#include "RenderableObjects\AnimatedModel\AnimatedInstanceModel.h"
#include "DebugInfo\DebugRender.h"
#include "DebugGUIManager.h"
#include "GameProcess.h"
#include "EngineProcess.h"
#include "Utils\Random.h"
#include "Characters\Player\Player.h"
#include "Cameras\ThPSCharacterCamera.h"
#include "PhysicController.h"
#include "ActionToInput.h"
#include "Helpers/MathHelper.h"
#include "Core.h"
#include "Base.h"

#if defined _DEBUG
#include "Memory\MemLeaks.h"
#endif

CPlayerTargetRunState::CPlayerTargetRunState( const std::string &_Name )
	: CState(_Name)
	, m_pTargetEnemy(NULL)
	, m_fMaxVelocityMovement(10.f)
	, m_fCurrentVelocityMovement(0.f)
	, m_fAccelerationStart(30.f)
	, m_fAccelerationEnd(65.f)
	, m_bStartState(true)
	, m_bEndState(false)
{

}

CPlayerTargetRunState::~CPlayerTargetRunState()
{

}

void CPlayerTargetRunState::OnEnter( CCharacter* _pCharacter )
{
#if defined _DEBUG
	if( CORE->IsDebugMode() )
	{
		CORE->GetDebugGUIManager()->GetDebugRender()->SetStateName("Target Run");
	}
#endif

	CGameProcess *l_pProcess	= static_cast<CGameProcess*>(CORE->GetProcess());
	m_pTargetEnemy				= l_pProcess->GetCharactersManager()->GetPreviewTargetEnemy();

	if( m_pTargetEnemy == NULL )
	{
		m_pTargetEnemy = static_cast<CPlayer*>(_pCharacter)->DetectEnemy();
	}

	l_pProcess->GetCharactersManager()->SetTargetEnemy( m_pTargetEnemy );

	m_fCurrentVelocityMovement	= 0.f;
	m_bStartState				= true;
	m_bEndState					= false;
}

void CPlayerTargetRunState::Execute( CCharacter* _pCharacter, float _fElapsedTime )
{
	CGameProcess *l_pProcess	= static_cast<CGameProcess*>(CORE->GetProcess());
	CPlayer *l_pPlayer			= static_cast<CPlayer*>(_pCharacter);
	CActionToInput *l_pInput	= CORE->GetActionToInput();

	if( !_pCharacter->GetLocked() )
	{
		if( m_pTargetEnemy != NULL )
		{
			if( !m_pTargetEnemy->IsAlive() )
			{
				l_pProcess->GetCharactersManager()->SetTargetEnemy( NULL );
				l_pPlayer->SetTargetFixed( false );
			}
		}

		//Mira si se desbloquea el target
		if( l_pInput->DoAction("PlayerTarget") )
		{
			l_pPlayer->SetTargetFixed( false );
		}

		//El jugador ataca
		if( l_pInput->DoAction("AttackPlayer") )
		{
			if( l_pPlayer->IsTargetFixed() )
			{
				_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("targetattack1") );
				_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animattack1") );
			}
			else
			{
				_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("attack1") );
				_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animattack1") );
			}
		}

		//Calcula el ángulo entre donde miras y el enemigo
		Vect3f l_PlayerDir	= l_pProcess->GetPlayerCamera()->GetDirection();
		Vect3f l_EnemyDir	= l_pProcess->GetCharactersManager()->GetTargetEnemy()->GetPosition() - _pCharacter->GetPosition();

		l_PlayerDir.y	= 0.0f;
		l_EnemyDir.y	= 0.0f;

		l_PlayerDir.Normalize(1.f);
		l_EnemyDir.Normalize(1.f);

		float l_fAngle	= l_EnemyDir.Dot( l_PlayerDir );
		l_fAngle		= mathUtils::ACos<float>( l_fAngle );

		if( mathUtils::Rad2Deg( l_fAngle ) > 0.1 )
		{
			_pCharacter->SetYaw( _pCharacter->GetYaw() - l_fAngle );
		}

		Vect3f l_Dir			= v3fZERO;
		bool l_bMovePlayer		= false;
		float l_fYawCharacter	= _pCharacter->GetYaw();

		//Se mueve el personaje con el target asignado
		if( l_pInput->DoAction("MovePlayerUp") )
		{
			l_Dir			= l_EnemyDir;
			l_bMovePlayer	= true;
		}
		else if( l_pInput->DoAction("MovePlayerDown") )
		{
			l_Dir			= -l_EnemyDir;
			l_bMovePlayer	= true;
		}
		else if( l_pInput->DoAction("MovePlayerLeft") )
		{
			float l_fYaw	= l_fYawCharacter + FLOAT_PI_VALUE / 2.f;
			l_Dir			= Vect3f( mathUtils::Cos<float>(l_fYaw), 0.f, mathUtils::Sin<float>(l_fYaw) );
			l_bMovePlayer	= true;
		}
		else if( l_pInput->DoAction("MovePlayerRight") )
		{
			_pCharacter->SetYaw( l_fYawCharacter + 2.f*l_fAngle);

			float l_fYaw	= _pCharacter->GetYaw() - FLOAT_PI_VALUE / 2.f;
			l_Dir			= Vect3f( mathUtils::Cos<float>(l_fYaw), 0.f, mathUtils::Sin<float>(l_fYaw) );
			l_bMovePlayer	= true;
		}

		_pCharacter->SetYaw( Helper::AngleFilter( _pCharacter->GetYaw() ) );

		if( l_bMovePlayer )
		{
			m_LastDirection = l_Dir;

			if( m_bEndState )
			{
				m_bStartState	= true;
				m_bEndState		= false;
			}
		}
		else
		{
			l_Dir = m_LastDirection;
			m_bEndState = true;
		}

		if( m_bStartState )
		{
			m_fCurrentVelocityMovement = m_fCurrentVelocityMovement + m_fAccelerationStart * _fElapsedTime;
			if( m_fCurrentVelocityMovement >= m_fMaxVelocityMovement )
			{
				m_fCurrentVelocityMovement = m_fMaxVelocityMovement;
				m_bStartState = false;
			}
		}
		else if( m_bEndState )
		{
			m_fCurrentVelocityMovement = m_fCurrentVelocityMovement - m_fAccelerationEnd * _fElapsedTime;
			if( m_fCurrentVelocityMovement <= 0.f )
			{
				m_fCurrentVelocityMovement = 0.f;
				m_bEndState = false;
			}
		}


		//Aplica la velocidad al movimiento
		l_Dir = l_Dir * m_fCurrentVelocityMovement * _fElapsedTime;

		//Mueve el controller físico
		_pCharacter->GetController()->Move( l_Dir, _fElapsedTime );

		//Pone la rotación del personaje
		_pCharacter->GetController()->SetYaw( l_fYawCharacter );
		_pCharacter->GetAnimatedModel()->SetYaw( -mathUtils::Rad2Deg( l_fYawCharacter) + 90.f );

		//Cambia el estado
		if( !l_bMovePlayer )
		{
			_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("idle") );
			_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animidle") );
		}
		else
		{
			//Mira si hace un salto
			if( l_pInput->DoAction("PlayerJump") )
			{
				_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("jump") );
				_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animjump") );
			}
			else if( !l_pPlayer->IsTargetFixed() )
			{
				_pCharacter->GetLogicFSM()->ChangeState( _pCharacter->GetLogicState("run") );
				_pCharacter->GetGraphicFSM()->ChangeState( _pCharacter->GetAnimationState("animrun") );

				l_pProcess->GetCharactersManager()->SetTargetEnemy( NULL );
			}
		}
	}
}

void CPlayerTargetRunState::OnExit( CCharacter* _pCharacter )
{
}

bool CPlayerTargetRunState::OnMessage( CCharacter* _pCharacter, const STelegram& _Message )
{
	if( _Message.Msg == Msg_Attack )
	{
		CRandom	l_Randomize;

		CCharacter *l_pEnemy	= static_cast<CGameProcess*>(CORE->GetProcess())->GetCharactersManager()->GetCharacterById(_Message.Sender);
		float l_fReceivedPain	= l_Randomize.getRandFloat( (float)(l_pEnemy->GetProperties()->GetStrong() / 2), (float)l_pEnemy->GetProperties()->GetStrong());
		float l_fPainToHit		= l_pEnemy->GetProperties()->GetStrong() * 0.95f;

		if( l_fReceivedPain >= l_fPainToHit )
		{
			static_cast<CPlayer*>(_pCharacter)->HitToPlayer();
		}

		return true;
	}
	else if( _Message.Msg == Msg_Push )
	{
		CCharacter *l_pEnemy	= static_cast<CGameProcess*>(CORE->GetProcess())->GetCharactersManager()->GetCharacterById(_Message.Sender);
		
		sDireccion * l_Info = (struct sDireccion *) _Message.ExtraInfo;
		_pCharacter->MoveTo2(l_Info->Direccion * 1.2f, l_Info->ElapsedTime);
		_pCharacter->GetSteeringEntity()->SetVelocity(Vect3f(0,0,0));
		LOGGER->AddNewLog(ELL_INFORMATION, "CPlayerTargetRunState::OnMessage -> PUSHED!!");
		return true;
	}

	return false;
}